Add API for quick tuning, otherwise improve the MIGraphX API

krzysz00 opened the pull request:
Note that this PR contains breaking changes (though, generally, minor ones) to MIGraphX, and the version define in the headers has been updated accordingly.

To summarize:
1. The tuning API has been split so that the full and quick tuning spaces are now separate objects with their own indexing, instead of having quick be a subset of full (which assumed that we'd have a sorting metric for the tuning space, which we don't). Therefore, mlirRockTuningParamGet has been renamed to mlirRockTuningParamGetFull and there is a mlirRockTuningParamGetQuick
2. The quick tuning space exists and is populated using our heuristic-mode logic, which uses a hard-coded list of configs per type and sorts them to prioritize kernels that don't require padding.
3. Both tuning spaces now filter out performance configurations that cannot result in valid kernels, but do **not** filter out every such config (they only remove those whose invalidity is clear from the parameters)
4. A new function has been added to construct the applicability pipeline, which is used to detect invalid performance configurations. When tuning, this should be run after applying a performance configuration to determine if that configuration is valid.
5. Various `int` parameters have been updated to more meaningful types.

Internal details for us:
1. The various PopulateParams classes have grown an isParamArtrPlausible() method that runs the AffixTuningParameters checks on a given perf config. This allows us to filter the tuning space.
2. The "prioritize paddilng-less kernels" sorter has now been made a public API so the tuning space constructrors can use it.
3. The InitParams* -> tuning parameters attribute constructor is now exposed for the non-xdlops case as well.
4. The tuning loop now iterates over these InitParams* structs, which are direct reflections of the perf config string.
5. Some of the C API tests have been updated, expaded, and, in one case, fixed.

krzysz00 committed:
Add API for quick tuning, otherwise improve the MIGraphX API

Note that this PR contains breaking changes (though, generally, minor
ones) to MIGraphX, and the version define in the headers has been
updated accordingly.

To summarize:
1. The tuning API has been split so that the full and quick tuning
spaces are now separate objects with their own indexing, instead of
having quick be a subset of full (which assumed that we'd have a
sorting metric for the tuning space, which we don't).
Therefore, mlirRockTuningParamGet has been renamed to
mlirRockTuningParamGetFull and there is a mlirRockTuningParamGetQuick
2. The quick tuning space exists and is populated using our
heuristic-mode logic, which uses a hard-coded list of configs per type
and sorts them to prioritize kernels that don't require padding.
3. Both tuning spaces now filter out performance configurations that
cannot result in valid kernels, but do **not** filter out every such
config (they only remove those whose invalidity is clear from the
parameters)
4. A new function has been added to construct the applicability
pipeline, which is used to detect invalid performance configurations.
When tuning, this should be run after applying a performance
configuration to determine if that configuration is valid.
5. Various `int` parameters have been updated to more meaningful types.

Internal details for us:
1. The various PopulateParams classes have grown an
isParamArtrPlausible() method that runs the AffixTuningParameters
checks on a given perf config. This allows us to filter the tuning
space.
2. The "prioritize paddilng-less kernels" sorter has now been made a
public API so the tuning space constructrors can use it.
3. The InitParams* -> tuning parameters attribute constructor is now
exposed for the non-xdlops case as well.
4. The tuning loop now iterates over these InitParams* structs, which
are direct reflections of the perf config string.
5. Some of the C API tests have been updated, expaded, and, in one
case, fixed.

krzysz00 commented:
MIGraphX folks, y'all're reviewers on this in order to detect reasons this API sucks.

manupak reviewed (approved):
Given the fact it is a sunk cost that we have the applicability pipeline (in a different universe, we could've done it via a partial lowering and op verifiers when constructing the tuning space but agreeing on that to be larger task altogether), I dont have any issues of using it now.

krzysz00 committed:
API reworks per Paul

1. The tuning database doesn't need to exist for looking up the
problem string anymore, because that was silly caching that wasn't
needed.
2. String-returning functions take a char* and a length of
caller-owned memory to copy the result in to, and return a size_t of
how many bytes they'd like to have copied.
2b. Constants for how big we think those buffers should be are defined.
3. Internally, a bunch of stuff now takes pointers to
maybe-stack-allocated buffers.
4. A bunch of uses of std::string (aka pointless heap allocations)
have been migrated to their LLVM equivalents. This also saves on
useless copies.
5. The problem string creator now returns LogicalResult() like we'd
want it to.

krzysz00 committed:
We wanted the .c_str() method to get the trailing null on

krzysz00 committed:
Fix typo causing test failure

manupak commented on mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp:
```diff
@@ -151,20 +188,29 @@ TunableParams *createTunableParamSpace(ModuleOp &mod) {
+                         ParamEntry *paramEntry) {
   // out of bound check.
-  if (pos < 0 || (unsigned int)pos > tuningSpace->tuningRange.size() - 1)
+  if (pos > tuningSpace->tuningRangeFull.size() - 1)
```
This seems not right...

krzysz00 replied:
The "buffer to be filled" thing was @pfultz2's request - I was going to go for `char **` arguments so we `malloc()` and return the buffer, but he was wanting to avoid the extra copy

krzysz00 replied:
... right, comment on the wrong thing.

This was a typo, fixed

manupak reviewed (commented):
There may be a typo...

I think what you've done additionally is passing a buffer to be filled with problemStr? 
which is good than piggybacking on the tuning table.

krzysz00 committed:
Merge branch 'develop' into migraphx-tuning-api-improvements

Also fix the typo

manupak reviewed (approved):
I dont see issues here....

krzysz00 committed:
Make the tuning table thread-safe

krzysz00 committed:
Fix size return values to include trailing null

krzysz00 committed:
Merge branch 'develop' into migraphx-tuning-api-improvements

krzysz00 committed:
Move our MIGraphX branch pointer so that we can land this

pfultz2 commented on mlir/include/mlir-c/Dialect/MIGraphX.h:
```diff
@@ -38,12 +38,27 @@ MLIR_CAPI_EXPORTED void mlirGetKernelAttrs(MlirModule module, uint32_t *attrs);
+/// and should, ideally, be called on a clone of the results of teh highlevel
+/// pipeline.
+MLIR_CAPI_EXPORTED void
+mlirMIGraphXAddApplicabilityPipeline(MlirPassManager pm);
```
Exactly how this supposed to be called? Wouldnt we add this always?

krzysz00 replied:
The applicability pipeline is a subset of the backend pipeline (that is, `backend(applicability(x)) == backend(x)`) such that failures during the applicability portion of the pipeline are failures of the request (ex. the tuning parameter ultimately made us use too much LDS) and not errors that need to be reported.

krzysz00 committed:
Merge branch 'develop' into migraphx-tuning-api-improvements

krzysz00 committed:
Merge branch 'develop' into migraphx-tuning-api-improvements

krzysz00 committed:
Make functions of string input take StringRef to make C++ interop easier

krzysz00 committed:
Merge branch 'develop' into migraphx-tuning-api-improvements

krzysz00 committed:
Make the tuning space kind an option to the tuning space constructor

Also add that infrastructure (couldBePerformant()) to enable a
filtered-down offline (full, as opposed to exhaustive) search space in
the future if we need one.

krzysz00 committed:
Merge branch 'develop' into migraphx-tuning-api-improvements
